File: web/src/lib/pricer.ts
// A TypeScript mirror of nitro's arbos/l2pricing/model.go and of
// internal/pricer/pricer.go, in integer basis points with BigInt so results
// match the Go pricer bit for bit, including the uint64 and int64 saturation.
// Never "simplify" this file to floating point.

import type { Constraint, LegacyParams } from "@/types";

export const ONE_IN_BIPS = 10_000n;
export const MAX_UINT64 = (1n << 64n) - 1n;
export const MAX_INT64 = (1n << 63n) - 1n;
export const MIN_INT64 = -(1n << 63n);

/** Go SaturatingUAdd: a + b for uint64, saturating at MaxUint64. */
export function saturatingUAdd(a: bigint, b: bigint): bigint {
  const sum = a + b;
  return sum > MAX_UINT64 ? MAX_UINT64 : sum;
}

/** Go SaturatingUSub: a - b for uint64, floored at zero. */
export function saturatingUSub(a: bigint, b: bigint): bigint {
  return b >= a ? 0n : a - b;
}

/** Go SaturatingUMul: a * b for uint64, saturating at MaxUint64. */
export function saturatingUMul(a: bigint, b: bigint): bigint {
  if (a === 0n || b === 0n) return 0n;
  const product = a * b;
  return product > MAX_UINT64 ? MAX_UINT64 : product;
}

/** Go saturatingCastToBips: a uint64 into int64 basis points, capped at MaxInt64. */
export function saturatingCastToBips(v: bigint): bigint {
  return v > MAX_INT64 ? MAX_INT64 : v;
}

/** Go NaturalToBips: v * 10_000 with saturating multiplication and cast. */
export function naturalToBips(v: bigint): bigint {
  return saturatingCastToBips(saturatingUMul(v, ONE_IN_BIPS));
}

/** Go SaturatingAddBips: int64 addition saturating at both bounds. */
export function saturatingAddBips(a: bigint, b: bigint): bigint {
  const sum = a + b;
  if (sum > MAX_INT64) return MAX_INT64;
  if (sum < MIN_INT64) return MIN_INT64;
  return sum;
}

/**
 * nitro ApproxExpBasisPoints: the degree `accuracy` Taylor polynomial of e^x,
 * evaluated with saturating unsigned arithmetic and integer division at every
 * step. Negative x takes the reciprocal branch, accuracy 0 is exactly 1.
 *   res = b + x/accuracy
 *   for i = accuracy-1 .. 1: res = b + res*x/(i*b)
 */
export function approxExpBips(x: bigint, accuracy = 4): bigint {
  if (accuracy === 0) return ONE_IN_BIPS;
  const negative = x < 0n;
  // Go negates in int64 and casts to uint64; -MinInt64 wraps to 2^63.
  const input = BigInt.asUintN(64, negative ? -x : x);
  const acc = BigInt(accuracy);
  const b = ONE_IN_BIPS;
  let res = b + input / acc;
  for (let i = acc - 1n; i > 0n; i--) {
    res = saturatingUAdd(b, saturatingUMul(res, input) / (i * b));
  }
  if (negative) return saturatingCastToBips((b * b) / res);
  return saturatingCastToBips(res);
}

export type ConstraintState = { target: bigint; window: bigint; backlog: bigint };

/** One constraint's exponent contribution in bips: NaturalToBips(backlog) / bips(window * target), as Go Step does. */
export function constraintExponentBips(c: ConstraintState): bigint {
  if (c.backlog <= 0n) return 0n;
  const divisor = saturatingUMul(c.window, c.target);
  if (divisor === 0n) return 0n;
  return naturalToBips(c.backlog) / saturatingCastToBips(divisor);
}

/** Sum of per-constraint contributions, each an integer division as in nitro, saturating. */
export function exponentBips(constraints: readonly ConstraintState[]): bigint {
  let total = 0n;
  for (const c of constraints) total = saturatingAddBips(total, constraintExponentBips(c));
  return total;
}

/** Base fee for an exponent: minBaseFee * P4(exponent) / 10000, or the floor when x <= 0. */
export function baseFeeFromExponent(minBaseFee: bigint, exponent: bigint): bigint {
  if (exponent <= 0n) return minBaseFee;
  return (minBaseFee * approxExpBips(exponent)) / ONE_IN_BIPS;
}

/** Multiplier over the floor in bips for an exponent (10000 = 1x). */
export function multiplierBips(exponent: bigint): bigint {
  return exponent <= 0n ? ONE_IN_BIPS : approxExpBips(exponent);
}

/** True e^x, for the comparison plot against P4. */
export function trueExpMultiplier(x: number): number {
  return Math.exp(x);
}

/** P4(x) evaluated through the integer pricer, as a float multiplier. */
export function p4Multiplier(x: number): number {
  const bips = BigInt(Math.round(x * 10_000));
  return Number(multiplierBips(bips)) / 10_000;
}

/** A JSON number into a uint64: truncated, floored at zero, capped at MaxUint64. */
export function toUint64(value: number): bigint {
  if (!Number.isFinite(value) || value <= 0) return 0n;
  const v = BigInt(Math.trunc(value));
  return v > MAX_UINT64 ? MAX_UINT64 : v;
}

/** API constraints (numbers) to pricer state (bigint). */
export function toState(constraints: readonly Pick<Constraint, "target" | "window" | "backlog">[]): ConstraintState[] {
  return constraints.map((c) => ({ target: toUint64(c.target), window: toUint64(c.window), backlog: toUint64(c.backlog) }));
}

/**
 * Client-side draining between ticks. Fractional dt is allowed here because
 * this only drives the gauge animation; each tick snaps back to sampled values.
 */
export function stepBacklogs(constraints: readonly Pick<Constraint, "target" | "backlog">[], dtSeconds: number): number[] {
  const dt = Math.max(0, dtSeconds);
  return constraints.map((c) => Math.max(0, c.backlog - c.target * dt));
}

/** Exponent contributions per constraint (integer bips) for a list of API constraints. */
export function contributionsBips(constraints: readonly Pick<Constraint, "target" | "window" | "backlog">[]): number[] {
  return toState(constraints).map((c) => Number(constraintExponentBips(c)));
}

/**
 * nitro Step: drain every backlog by dt * target, then compute the exponent and fee.
 * Mutates `constraints`. Returns the fee that applies to the block being opened.
 */
export function step(
  constraints: ConstraintState[],
  dtSeconds: bigint,
  minBaseFee: bigint,
): { baseFee: bigint; exponent: bigint; contributions: bigint[] } {
  const contributions: bigint[] = [];
  let exponent = 0n;
  for (const c of constraints) {
    c.backlog = saturatingUSub(c.backlog, saturatingUMul(dtSeconds, c.target));
    const x = constraintExponentBips(c);
    contributions.push(x);
    exponent = saturatingAddBips(exponent, x);
  }
  return { baseFee: baseFeeFromExponent(minBaseFee, exponent), exponent, contributions };
}

/** nitro AddGas: every constraint absorbs every unit of gas, saturating. Mutates `constraints`. */
export function addGas(constraints: ConstraintState[], gasUsed: bigint): void {
  for (const c of constraints) c.backlog = saturatingUAdd(c.backlog, gasUsed);
}

export type LegacyState = { speedLimit: bigint; inertia: bigint; tolerance: bigint; backlog: bigint };

export function toLegacyState(legacy: Pick<LegacyParams, "speedLimit" | "inertia" | "tolerance" | "backlog">): LegacyState {
  return {
    speedLimit: toUint64(legacy.speedLimit),
    inertia: toUint64(legacy.inertia),
    tolerance: toUint64(legacy.tolerance),
    backlog: toUint64(legacy.backlog),
  };
}

/**
 * Legacy exponent: bips(backlog - tolerance*speedLimit) / (inertia*speedLimit)
 * when above tolerance. Mirrors nitro's updatePricingModelLegacy and the Go
 * pricer exactly: the tolerance threshold is a plain uint64 multiply that
 * wraps on overflow, the excess is cast to bips saturating, and the inertia
 * denominator is a saturating multiply cast to bips saturating. A zero
 * denominator (nitro would panic) yields no exponent.
 */
export function legacyExponentBips(s: LegacyState): bigint {
  const threshold = BigInt.asUintN(64, s.tolerance * s.speedLimit);
  if (s.backlog <= threshold) return 0n;
  const inertia = saturatingCastToBips(saturatingUMul(s.inertia, s.speedLimit));
  if (inertia <= 0n) return 0n;
  return naturalToBips(s.backlog - threshold) / inertia;
}

/** Legacy Step: drain by dt * speedLimit, then price. Mutates `s`. */
export function legacyStep(s: LegacyState, dtSeconds: bigint, minBaseFee: bigint): { baseFee: bigint; exponent: bigint } {
  s.backlog = saturatingUSub(s.backlog, saturatingUMul(dtSeconds, s.speedLimit));
  const exponent = legacyExponentBips(s);
  return { baseFee: baseFeeFromExponent(minBaseFee, exponent), exponent };
}

export function legacyAddGas(s: LegacyState, gasUsed: bigint): void {
  s.backlog = saturatingUAdd(s.backlog, gasUsed);
}

File: internal/pricer/pricer.go
// Package pricer is a pure re-implementation of the Arbitrum Nitro L2 gas pricing model
// (arbos/l2pricing/model.go). It uses the same integer basis point arithmetic as nitro, including
// saturation, so a replay over block headers reproduces the chain's base fee bit for bit.
//
// Two models are supported: the multi-constraint model introduced with ArbOS 50 (a list of
// {target, window, backlog} triples) and the legacy single backlog model.
package pricer

import (
	"math"
	"math/big"
)

// Bips is a value in basis points (1/10_000), nitro's arbmath.Bips.
type Bips int64

// OneInBips is 100% expressed in basis points.
const OneInBips Bips = 10_000

// InitialMinimumBaseFeeWei is nitro's genesis minimum base fee (0.1 gwei), in force until the first
// recorded setMinimumL2BaseFee owner action.
const InitialMinimumBaseFeeWei = 100_000_000

// Constraint is one gas pricing constraint: a gas target per second, an adjustment window in seconds,
// and the current backlog in gas.
type Constraint struct {
	Target  uint64
	Window  uint64
	Backlog uint64
}

// Legacy holds the pre-constraint pricer parameters.
type Legacy struct {
	SpeedLimit uint64
	Inertia    uint64
	Tolerance  uint64
	Backlog    uint64
}

// State is the mutable pricer state. When Constraints is non-empty the
// multi-constraint model is used, otherwise Legacy.
type State struct {
	Constraints []Constraint
	Legacy      *Legacy
	MinBaseFee  *big.Int
}

func (s *State) IsLegacy() bool {
	return len(s.Constraints) == 0
}

func (s *State) Clone() *State {
	out := &State{MinBaseFee: new(big.Int)}
	if s.MinBaseFee != nil {
		out.MinBaseFee.Set(s.MinBaseFee)
	}
	if len(s.Constraints) > 0 {
		out.Constraints = make([]Constraint, len(s.Constraints))
		copy(out.Constraints, s.Constraints)
	}
	if s.Legacy != nil {
		l := *s.Legacy
		out.Legacy = &l
	}
	return out
}

// Backlogs returns the current backlogs, one per constraint (one element for the legacy model).
func (s *State) Backlogs() []uint64 {
	if s.IsLegacy() {
		if s.Legacy == nil {
			return nil
		}
		return []uint64{s.Legacy.Backlog}
	}
	out := make([]uint64, len(s.Constraints))
	for i, c := range s.Constraints {
		out[i] = c.Backlog
	}
	return out
}

// SetBacklogs overwrites the backlogs. Extra values are ignored and missing ones leave the existing
// backlog in place.
func (s *State) SetBacklogs(backlogs []uint64) {
	if s.IsLegacy() {
		if s.Legacy != nil && len(backlogs) > 0 {
			s.Legacy.Backlog = backlogs[0]
		}
		return
	}
	for i := range s.Constraints {
		if i < len(backlogs) {
			s.Constraints[i].Backlog = backlogs[i]
		}
	}
}

// AddGas adds gas used by a transaction to every constraint backlog, saturating.
func (s *State) AddGas(gas uint64) {
	if s.IsLegacy() {
		if s.Legacy != nil {
			s.Legacy.Backlog = SaturatingUAdd(s.Legacy.Backlog, gas)
		}
		return
	}
	for i := range s.Constraints {
		s.Constraints[i].Backlog = SaturatingUAdd(s.Constraints[i].Backlog, gas)
	}
}

// Step advances the pricer by dt seconds as ArbOS does at the start of a block: backlogs are paid down
// at the target rate, then the exponent and base fee are computed. It returns the base fee, the total
// exponent and the per-constraint exponents.
func (s *State) Step(dt uint64) (baseFee *big.Int, exponent Bips, perConstraint []Bips) {
	minBaseFee := s.MinBaseFee
	if minBaseFee == nil {
		minBaseFee = new(big.Int)
	}
	if s.IsLegacy() {
		return s.stepLegacy(dt, minBaseFee)
	}
	perConstraint = make([]Bips, len(s.Constraints))
	for i := range s.Constraints {
		c := &s.Constraints[i]
		c.Backlog = SaturatingUSub(c.Backlog, SaturatingUMul(dt, c.Target))
		if c.Backlog == 0 {
			continue
		}
		divisor := SaturatingUMul(c.Window, c.Target)
		if divisor == 0 {
			continue
		}
		perConstraint[i] = NaturalToBips(c.Backlog) / saturatingCastToBips(divisor)
		exponent = SaturatingAddBips(exponent, perConstraint[i])
	}
	return BaseFeeFromExponent(minBaseFee, exponent), exponent, perConstraint
}

func (s *State) stepLegacy(dt uint64, minBaseFee *big.Int) (baseFee *big.Int, exponent Bips, perConstraint []Bips) {
	if s.Legacy == nil {
		return new(big.Int).Set(minBaseFee), 0, nil
	}
	l := s.Legacy
	l.Backlog = SaturatingUSub(l.Backlog, SaturatingUMul(dt, l.SpeedLimit))
	// Mirrors nitro's updatePricingModelLegacy exactly: the tolerance threshold is a plain uint64
	// multiply (it wraps on overflow), the excess is cast to int64 saturating, and the inertia
	// denominator is a saturating multiply cast to Bips saturating. Nitro would panic on a zero
	// denominator; that case yields no exponent here.
	threshold := l.Tolerance * l.SpeedLimit
	if l.Backlog > threshold {
		inertia := saturatingCastToBips(SaturatingUMul(l.Inertia, l.SpeedLimit))
		if inertia > 0 {
			exponent = NaturalToBips(l.Backlog-threshold) / inertia
		}
	}
	return BaseFeeFromExponent(minBaseFee, exponent), exponent, []Bips{exponent}
}

// BaseFeeFromExponent is minBaseFee * approxExp(exponent) / 10_000 for a positive exponent and
// minBaseFee otherwise (nitro's BigMulByBips).
func BaseFeeFromExponent(minBaseFee *big.Int, exponent Bips) *big.Int {
	if exponent <= 0 {
		return new(big.Int).Set(minBaseFee)
	}
	mult := ApproxExpBips(exponent, 4)
	out := new(big.Int).Mul(minBaseFee, big.NewInt(int64(mult)))
	return out.Div(out, big.NewInt(int64(OneInBips)))
}

// ApproxExpBips is nitro's arbmath.ApproxExpBasisPoints: a degree `accuracy` Taylor polynomial of exp
// evaluated in basis points with saturating unsigned arithmetic. ApproxExpBips(0, n) == OneInBips.
func ApproxExpBips(x Bips, accuracy uint64) Bips {
	if accuracy == 0 {
		return OneInBips
	}
	negative := x < 0
	if negative {
		x = -x
	}
	input := uint64(x)
	b := uint64(OneInBips)
	res := b + input/accuracy
	for i := accuracy - 1; i > 0; i-- {
		res = SaturatingUAdd(b, SaturatingUMul(res, input)/(i*b))
	}
	if negative {
		return saturatingCastToBips(b * b / res)
	}
	return saturatingCastToBips(res)
}

// NaturalToBips converts a natural number into basis points, saturating at MaxInt64 (nitro's
// NaturalToBips/SaturatingCastToBips).
func NaturalToBips(v uint64) Bips {
	return saturatingCastToBips(SaturatingUMul(v, uint64(OneInBips)))
}

func saturatingCastToBips(v uint64) Bips {
	if v > math.MaxInt64 {
		return Bips(math.MaxInt64)
	}
	return Bips(v)
}

func SaturatingAddBips(a, b Bips) Bips {
	sum := a + b
	if b > 0 && sum < a {
		return Bips(math.MaxInt64)
	}
	if b < 0 && sum > a {
		return Bips(math.MinInt64)
	}
	return sum
}

func SaturatingUAdd(a, b uint64) uint64 {
	sum := a + b
	if sum < a {
		return math.MaxUint64
	}
	return sum
}

func SaturatingUSub(a, b uint64) uint64 {
	if b >= a {
		return 0
	}
	return a - b
}

func SaturatingUMul(a, b uint64) uint64 {
	if a == 0 || b == 0 {
		return 0
	}
	if a > math.MaxUint64/b {
		return math.MaxUint64
	}
	return a * b
}
